src/backend/drm/dumb.rs

```diff
@@ -0,0 +1,185 @@
+        }
+    }
+}
+
+impl AsRef<framebuffer::Handle> for DumbFramebuffer {
+    fn as_ref(&self) -> &framebuffer::Handle {
+        &self.fb
+    }
+}
+
+impl Framebuffer for DumbFramebuffer {
+    fn format(&self) -> drm_fourcc::DrmFormat {
+        self.format
+    }
+}
+
+struct PlanarDumbBuffer<'a>(&'a DumbBuffer);
+
+impl<'a> Buffer for PlanarDumbBuffer<'a> {
+    fn size(&self) -> Size<i32, BufferCoords> {
+        self.0.size()
+    }
+
+    fn format(&self) -> DrmFormat {
+        self.0.format()
+    }
+}
+
+impl<'a> PlanarBuffer for PlanarDumbBuffer<'a> {
+    fn size(&self) -> (u32, u32) {
+        let size = self.0.size();
+        (size.w as u32, size.h as u32)
+    }
+
+    fn format(&self) -> Fourcc {
+        self.0.format().code
+    }
+
+    fn modifier(&self) -> Option<DrmModifier> {
+        Some(self.0.format().modifier)
+    }
+
+    fn pitches(&self) -> [u32; 4] {
+        [self.0.handle().pitch(), 0, 0, 0]
+    }
+
+    fn handles(&self) -> [Option<drm::buffer::Handle>; 4] {
+        [Some(self.0.handle().handle()), None, None, None]
+    }
+
+    fn offsets(&self) -> [u32; 4] {
+        [0, 0, 0, 0]
+    }
+}
+
+struct OpaqueBufferWrapper<'a, B>(&'a B);
+impl<'a, B> PlanarBuffer for OpaqueBufferWrapper<'a, B>
+where
+    B: PlanarBuffer,
+{
+    fn size(&self) -> (u32, u32) {
+        self.0.size()
+    }
+
+    fn format(&self) -> Fourcc {
+        let fmt = self.0.format();
+        get_opaque(fmt).unwrap_or(fmt)
+    }
+
+    fn modifier(&self) -> Option<DrmModifier> {
+        self.0.modifier()
+    }
+
+    fn pitches(&self) -> [u32; 4] {
+        self.0.pitches()
+    }
+
+    fn handles(&self) -> [Option<drm::buffer::Handle>; 4] {
+        self.0.handles()
+    }
+
+    fn offsets(&self) -> [u32; 4] {
+        self.0.offsets()
+    }
+}
+
+/// Attach a [`framebuffer::Handle`] to an [`DumbBuffer`]
+#[profiling::function]
+pub fn framebuffer_from_dumb_buffer(
+    drm: &DrmDeviceFd,
+    buffer: &DumbBuffer,
+    use_opaque: bool,
+) -> Result<DumbFramebuffer, AccessError> {
+    let buffer = PlanarDumbBuffer(buffer);
+
+    let format = Buffer::format(&buffer);
+    let modifier = buffer.modifier();
+    let flags = if modifier.is_some() {
+        FbCmd2Flags::MODIFIERS
+    } else {
+        FbCmd2Flags::empty()
+    };
+
+    let ret = if use_opaque {
+        let opaque_wrapper = OpaqueBufferWrapper(&buffer);
+        drm.add_planar_framebuffer(&opaque_wrapper, flags).map(|fb| {
+            (
+                fb,
+                drm_fourcc::DrmFormat {
+                    code: opaque_wrapper.format(),
+                    modifier: modifier.unwrap_or(DrmModifier::Invalid),
+                },
+            )
+        })
+    } else {
+        drm.add_planar_framebuffer(&buffer, flags).map(|fb| (fb, format))
+    };
+
+    let (fb, format) = match ret {
+        Ok(fb) => fb,
+        Err(source) => {
+            let fourcc = format.code;
```

Comment:
I think it'd be worth putting a debug!() here since most/all drivers we use should support addfb2, and this fallback *shouldn't* happen

Reply:
You mean a log message? So more like a `tracing::warn`?

Reply:
If a warn!, we should make sure to not send this on every occurrence, but once.

Reply:
Added a `warn` wrapped in an `Once::call_once` to warn once about the legacy fallback path